### PR TITLE
Bcf: Fix sensor shell example

### PR DESCRIPTION
**Pull request by Ayush1325**
The following commits adds support for `samples/sensor/sensor_shell` for beagleconnect freedom.

Currently, it seems `hdc2010-humidity` sensor does not show up on `sensor info` command but works fine with `sensor get` so help on that will be appreciated.

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
I think it would make a lot more sense if this was written as a generic "GPIO controlled I2C bus switch" rather than pointing out a specific general purpose analog switch IC.

How about rewording the whole thing so it sounds generic?

**Comment by Ayush1325** [comment]
> I think it would make a lot more sense if this was written as a generic "GPIO controlled I2C bus switch" rather than pointing out a specific general purpose analog switch IC.
> 
> How about rewording the whole thing so it sounds generic?

Just to confirm you are talking about the driver right (not the dt-bindings)?

**Comment by fabiobaltieri** [comment]
> Just to confirm you are talking about the driver right (not the dt-bindings)?

Well the whole thing, binding name and compatible, driver name, function prefixes, board definition etc. I the board dts should not have been submitted as is in the first place - we normally don't take trees with off-tree bindings but this slipped through I guess.

**Comment by Ayush1325** [comment]
> Well the whole thing, binding name and compatible, driver name, function prefixes, board definition etc. I the board dts should not have been submitted as is in the first place - we normally don't take trees with off-tree bindings but this slipped through I guess.

Well, `ti,ts5a2066` compatible string is already present in the bcf dts (which is definitely weird), so not sure whether it is worth changing that. 
I guess Zephyr doesn't follow the whole bindings should describe the actual hardware thing from the Linux kernel? 
I am happy to make any changes that will suit upstream zephyr. I would just prefer to fix the half broken state of upstream support for bcf.

**Comment by fabiobaltieri** [comment]
> Well, `ti,ts5a2066` compatible string is already present in the bcf dts (which is definitely weird), so not sure whether it is worth changing that.
> I guess Zephyr doesn't follow the whole bindings should describe the actual hardware thing from the Linux kernel?

Yeah I think it should be changed, you could add multiple compatibles if you really want to but I don't really see a point, it's a generic analog switch, what's the point of having a driver referring to it by part number? It would still describe the actual hardware.

**Comment by fabiobaltieri** [comment]
> Well, the linux kernel actually does pretty much require the part thing (even in cases where hardware is not important is running some sort of special firmware), but I digress.

At some extent, it's not like you'd expect stuff like a `gpio-keys` entry to have the compatible for the exact switch p/n for example. But to your point, it's certainly more common in the Linux case, mainly because there ther's a solid infrastructure to handle multiple compatibles for the same drivers and introduce changes in behavior down the road, that's unlikely be the case for something as simple as this though. You have a more specific example to look for? Happy to dig in that direction if you feel like.

> I am not familiar enough to say if there is or will be other parts that work exactly the same as this, but I am happy to make the changes if that's what upstream expects.

Dual channel SPST analog switches? I would lay towards a ["yes"](https://www.mouser.ie/c/semiconductors/switch-ics/analog-switch-ics/?q=analog%20switch&configuration=2%20x%20SPST&number%20of%20channels=2%20Channel). :-)

**Comment by Ayush1325** [comment]
> Dual channel SPST analog switches? I would lay towards a ["yes"](https://www.mouser.ie/c/semiconductors/switch-ics/analog-switch-ics/?q=analog%20switch&configuration=2%20x%20SPST&number%20of%20channels=2%20Channel). :-)

Ok. does`dual_chan_spst_analog_switch`  seem fine for all terminology? Or maybe `dchan_spst_aswitch`?

**Comment by fabiobaltieri** [comment]
> Ok. does`dual_chan_spst_analog_switch` seem fine for all terminology? Or maybe `dchan_spst_aswitch`?

An analog switch can be used for many different things, your driver is specifically about switching an i2c line using a gpio based switch regardless of how the thing is implemented so I would go with something like `gpio-i2c-switch`.  @teburd any opinions about this?

File: boards/arm/beagle_bcf/Kconfig.defconfig [hunk 1]
```diff
@@ -9,4 +9,14 @@ if BOARD_BEAGLECONNECT_FREEDOM
 config BOARD
 	default "beagleconnect_freedom"
 
+if SENSOR
+
+config OPT3001
+	default y
+
+config TI_HDC20XX
+	default y
```
Comment (fabiobaltieri) [inline boards/arm/beagle_bcf/Kconfig.defconfig:25]: neither of these should be needed, they are enabled automatically once the corresponding devicetree entry is instantiated, this bit:

```
        default y
        depends on DT_HAS_TI_OPT3001_ENABLED
```

Can you drop them?

File: drivers/i2c/gpio_i2c_switch.c [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+#define DEFINE_GPIO_I2C_SWITCH(inst)                                                               \
+                                                                                                   \
+	static struct gpio_i2c_switch_data gpio_i2c_switch_dev_data_##inst;                        \
+                                                                                                   \
+	static const struct gpio_i2c_switch_config gpio_i2c_switch_dev_cfg_##inst = {              \
+		.bus = DEVICE_DT_GET(DT_PHANDLE(DT_INST(inst, DT_DRV_COMPAT), controller)),        \
+		.gpiodev = DEVICE_DT_GET(DT_PHANDLE(DT_INST(inst, DT_DRV_COMPAT), gpios)),         \
+		.gpiopin = DT_INST_GPIO_PIN(inst, gpios),                                          \
+		.gpioflags = DT_INST_GPIO_FLAGS(inst, gpios),                                      \
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:102]: Use `GPIO_DT_SPEC_GET` here.

File: drivers/i2c/gpio_i2c_switch.c [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	const struct gpio_i2c_switch_config *config = dev->config;
+	int res;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+	if (res != 0) {
+		return res;
+	}
+
+	gpio_i2c_switch_enable(dev);
+	k_busy_wait(100);
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:71]: add a define for that 100

File: drivers/i2c/gpio_i2c_switch.c [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = gpio_pin_configure(config->gpiodev, config->gpiopin, GPIO_OUTPUT_HIGH);
+
+	return res;
+}
+
+static int gpio_i2c_switch_disable(const struct device *dev)
+{
+	int res = 0;
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:50]: no need to initialize "res", you set it immediately just two lines after

File: drivers/i2c/gpio_i2c_switch.c [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+int gpio_i2c_switch_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return i2c_configure(config->bus, dev_config);
+}
+
+static int gpio_i2c_switch_enable(const struct device *dev)
+{
+	int res = 0;
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:40]: no need to initialize res

**Review by fabiobaltieri — COMMENTED** [review]
Cool,few comments, from the design standpoint, I'd be tempted to suggest that this should be `on-bus: i2c` and be instantiated as a child of the i2c device. That would place it in the right point in the i2c hierarchy, the only problem is that doing that would force you to assign a dummy address to the device, so I'm not too sure about it. @teburd what do you think?

File: drivers/i2c/gpio_i2c_switch.c [hunk 6]
```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:2]: also 2023 :-)

**Comment by Ayush1325** [comment]
Can anyone tell me why Footprint Delta is failing?

**Comment by yperess** [comment]
> Can anyone tell me why Footprint Delta is failing?

I'm re-running with debug logging

File: drivers/i2c/gpio_i2c_switch.c [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	.transfer = gpio_i2c_switch_transfer,
+};
+
+static int gpio_i2c_switch_init(const struct device *dev)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	gpio_pin_configure(config->gpio.port, config->gpio.pin, GPIO_OUTPUT_LOW);
+
+	return 0;
```
Comment (yperess) [inline drivers/i2c/gpio_i2c_switch.c:94]: Maybe return the result from `gpio_pin_configure`?

File: drivers/i2c/gpio_i2c_switch.c [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+static int gpio_i2c_switch_transfer(const struct device *dev, struct i2c_msg *msgs,
+				    uint8_t num_msgs, uint16_t addr)
+{
+	int res;
+	struct gpio_i2c_switch_data *data = dev->data;
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+	if (res) {
```
Comment (yperess) [inline drivers/i2c/gpio_i2c_switch.c:69]: Please be explicit, `if (res != 0)`

File: drivers/i2c/gpio_i2c_switch.c [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+static int gpio_i2c_switch_transfer(const struct device *dev, struct i2c_msg *msgs,
+				    uint8_t num_msgs, uint16_t addr)
+{
+	int res;
+	struct gpio_i2c_switch_data *data = dev->data;
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
```
Comment (yperess) [inline drivers/i2c/gpio_i2c_switch.c:68]: That's a pretty long timeout, can you make it configurable via the Kconfig maybe?

**Review by yperess — CHANGES_REQUESTED** [review]

File: drivers/i2c/Kconfig [hunk 10]
```diff
@@ -195,4 +195,9 @@ config I2C_RV32M1_LPI2C
 	help
 	  Enable the RV32M1 LPI2C driver.
 
+config GPIO_I2C_SWITCH
+	bool "GPIO controlled I2C bus switch"
```
Comment (fabiobaltieri) [inline drivers/i2c/Kconfig:199]:
```
depends on DT_HAS_GPIO_I2C_SWITCH_ENABLED
default y
```

and then drop the change in the board file

File: drivers/i2c/gpio_i2c_switch.c [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+ * Copyright (c) 2020 Innoseis BV
+ *
+ * Based on i2c_tca9656a.c
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_i2c_switch
+
+#define GPIO_I2C_TRANSFER_DELAY 100
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:13]: use a suffix for time units, `GPIO_I2C_TRANSFER_DELAY_US`

File: drivers/i2c/gpio_i2c_switch.c [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+static int gpio_i2c_switch_transfer(const struct device *dev, struct i2c_msg *msgs,
+				    uint8_t num_msgs, uint16_t addr)
+{
+	int res;
+	struct gpio_i2c_switch_data *data = dev->data;
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+	if (res) {
+		return res;
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:70]: you are missing an `unlock` in this branch

File: drivers/i2c/gpio_i2c_switch.c [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	return 0;
+}
+
+#define DEFINE_GPIO_I2C_SWITCH(inst)                                                               \
+                                                                                                   \
+	static struct gpio_i2c_switch_data gpio_i2c_switch_dev_data_##inst;                        \
+                                                                                                   \
+	static const struct gpio_i2c_switch_config gpio_i2c_switch_dev_cfg_##inst = {              \
+		.bus = DEVICE_DT_GET(DT_PHANDLE(DT_DRV_INST(inst), controller)),        \
+		.gpio = GPIO_DT_SPEC_GET(DT_DRV_INST(inst), gpios),         \
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:103]: can you align the `\`?

File: drivers/i2c/gpio_i2c_switch.c [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	gpio_i2c_switch_enable(dev);
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY);
+	res = i2c_transfer(config->bus, msgs, num_msgs, addr);
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY);
+	gpio_i2c_switch_disable(dev);
+
+	k_mutex_unlock(&data->lock);
+	return res;
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:80]: Either add a label and goto from the error condition above, or change this to `return 0;`, in any case handle the return value close to where it's set, that is check for it after `i2c_transfer`

File: drivers/i2c/gpio_i2c_switch.c (hunk 12 above)
Comment (Ayush1325) [reply drivers/i2c/gpio_i2c_switch.c:63]: Why would an unlock be needed if the mutex fails to lock?

**Review by yperess — DISMISSED** [review]

File: drivers/i2c/gpio_i2c_switch.c (hunk 12 above)
Comment (fabiobaltieri) [reply drivers/i2c/gpio_i2c_switch.c:63]: oh, sorry, misread the code - yes you are right

File: drivers/i2c/gpio_i2c_switch.c [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+const struct i2c_driver_api gpio_i2c_switch_api_funcs = {
+	.configure = gpio_i2c_switch_configure,
+	.transfer = gpio_i2c_switch_transfer,
+};
+
+static int gpio_i2c_switch_init(const struct device *dev)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return gpio_pin_configure(config->gpio.port, config->gpio.pin, GPIO_OUTPUT_LOW);
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:89]: use `gpio_pin_configure_dt`

File: drivers/i2c/gpio_i2c_switch.c [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+const struct i2c_driver_api gpio_i2c_switch_api_funcs = {
+	.configure = gpio_i2c_switch_configure,
+	.transfer = gpio_i2c_switch_transfer,
+};
+
+static int gpio_i2c_switch_init(const struct device *dev)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:88]: I think you are missing a `k_mutex_init` for the lock here

File: drivers/i2c/gpio_i2c_switch.c [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+		return res;
+	}
+
+	gpio_i2c_switch_enable(dev);
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY_US);
+	res = i2c_transfer(config->bus, msgs, num_msgs, addr);
+	if (res != 0) {
+		goto disable_switch;
+	}
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY_US);
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:72]: hey thinking about the logic, is this wait needed? I understand the one before the transaction if the switch has any settling time, but then why would you wait BEFORE setting it back? This should be placed after `gpio_i2c_switch_disable`.

File: drivers/i2c/gpio_i2c_switch.c [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return i2c_configure(config->bus, dev_config);
+}
+
+static int gpio_i2c_switch_enable(const struct device *dev)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return gpio_pin_configure(config->gpio.port, config->gpio.pin, GPIO_OUTPUT_HIGH);
```
Comment (fabiobaltieri) [inline drivers/i2c/gpio_i2c_switch.c:44]: `gpio_pin_configure_dt`, to be honest at this point you may as well get rid of the function and call this from `gpio_i2c_switch_transfer` directly

File: drivers/i2c/gpio_i2c_switch.c (hunk 9 above)
Comment (Ayush1325) [reply drivers/i2c/gpio_i2c_switch.c]: I talked to the orignial authors and it seems the reason for the abnormally long timeout was that they weren't sure about the other delays at the time. I have now made it dependent on the transfer delay (2 * delay + 100us). Should it still be configurable?

**Review by yperess — DISMISSED** [review]

File: drivers/i2c/gpio_i2c_switch.c (hunk 17 above)
Comment (Ayush1325) [reply drivers/i2c/gpio_i2c_switch.c]: Yes, it can be removed. I have also reduced the gpio toggle delay to 1 us after checking the [data sheet](https://www.ti.com/lit/ds/symlink/ts5a2066.pdf?ts=1699344464626&ref_url=https%253A%252F%252Fwww.google.com%252F) according to which the max on/off time is 12.9 ns.

**Review by teburd — APPROVED** [review]

**Review by vaishnavachath — APPROVED** [review]

File: drivers/i2c/gpio_i2c_switch.c (hunk 17 above)
Comment (jadonk) [reply drivers/i2c/gpio_i2c_switch.c]: I believe there was a delay in case there was any propagation time through the controller. The intent was to avoid cutting the bus off before the transaction is completed at the hardware. As long as we are past the ACK, all should be good and no wait after the transaction *before* flipping the I/O is needed. And, as it seems, the delay time from flipping the I/O and having it propagate to the switch before new I2C transfers can happen seems negligable. Thanks for going through all this effort to clean up the hacks and sorry it slipped through the initial upstream submission.

File: drivers/i2c/gpio_i2c_switch.c (hunk 9 above)
Comment (jadonk) [reply drivers/i2c/gpio_i2c_switch.c]: I think the dependency is on the MCU and switch, so a configuration option makes sense with a default definition provided by the platform. For the base configuration, it should be provided and not given a default as it is platform usage dependent.

File: drivers/i2c/gpio_i2c_switch.c (hunk 17 above)
Comment (fabiobaltieri) [reply drivers/i2c/gpio_i2c_switch.c]: Of cousre, it's a good addition, sure there's more board around that could need this. Also now that I look at it again I realized I missed a misuse of the GPIO API :-) let me put together a quick followup for that.
